osidb/filters.py

```diff
@@ -116,21 +115,13 @@ def changed_after_filter(self, queryset, name, value):
         """
         Returns a Flaw if it or any of its affects/trackers have been updated after `value`
         """
-        return queryset.filter(
-            Q(updated_dt__gte=value)
-            | Q(affects__updated_dt__gte=value)
-            | Q(affects__trackers__updated_dt__gte=value)
-        ).distinct()
+        return queryset.filter(local_updated_dt__gte=value)
 
     def changed_before_filter(self, queryset, name, value):
         """
         Returns a Flaw if it or any of its affects/trackers have been updated before `value`
         """
-        return queryset.filter(
-            Q(updated_dt__lte=value)
-            | Q(affects__updated_dt__lte=value)
-            | Q(affects__trackers__updated_dt__lte=value)
-        ).distinct()
+        return queryset.filter(local_updated_dt__lte=value)
```

Comment:
This is a time/space optimisation eg. push processing time at ingestion time to avoid spending resolving **OR** during runtime.

Were you able to look at _.query / .explain()_ for current **queryset.filter** and confirm it is the **OR** condition slowing this down. 

Does current queryset.filter hit indexes (they exist right ?) for resolving comparison on **affects__updated_dt__lte** ? If they do not then I would suggest doing that optimisation first (eg. ensure index exists and are used) before 'baking in' as proposed in this PR.

Reply:
I can take a look but it seemed pretty obvious to me, we're doing a lot of potentially useless joins and on top of it we add a distinct, just to determine whether the data has changed, seems like a clear case for denormalization.

I will take a look, but by "existing indexes are not being used" do you mean by the Django ORM / Programmer error or by the Postgres query planner?

Reply:
To determine root cause of bad performance - its always good to view query plan ... SQL with a lot of joins can often 'boil down' to an efficient query plan ... but agree yes we should strive to write efficient SQL queries to begin with. 

In this specific case I feel like with the very little data we have that we should have good enough performance here (even with OR queries) using updated_dt indexes.

Using .explain() in django or via pgcli EXPLAIN ANALYZE on representative SQL query (using those updated_dt) should show if indexes are being used or not ... 

If pg_stat works we could do
```
SELECT schemaname, pg_stat_user_indexes.indexrelname, idx_scan, idx_tup_read, idx_tup_fetch                    
FROM pg_stat_user_indexes                                                                                      
WHERE pg_stat_user_indexes.indexrelname = '$1'                                                                 
```
where put index name for $1 to show usage of related indexes.

If indexes are not being used then that is a candidate for slow query and maybe we should understand why the query planner does not choose to use them ... if they are being used then your PR addresses it nicely ;)

I can also check tomorrow ....